picas/actors.py

```diff
@@ -5,91 +5,221 @@
+        state (default: 'todo' state).
+
+        @param signum: signal to listen to and act upon
+        @param frame: stack frame, defaults to None, see https://docs.python.org/3/library/signal.html#signal.signal
+        """
+        log.info(f'PiCaS shutting down, called with signal {signum}')
+
+        # gracefully kill the process running token code, it needs to stop before we update the token state
+        if self.subprocess and self.subprocess.poll() is None:
+            log.info('Terminating execution of token')
+            self.subprocess.terminate()
+            try:
+                self.subprocess.communicate(timeout=30)  # wait 30 seconds for termination, value chosen to allow complex processes to stop
+            except subprocess.TimeoutExpired:
+                log.info('Killing subprocess')
+                self.subprocess.kill()
+                self.subprocess.communicate()
+
+        # update the token state, if reset vaue is None, do nothing.
+        if self.current_task and self.token_reset_values is not None:
+            self.current_task['lock'] = self.token_reset_values[0]
+            self.current_task['done'] = self.token_reset_values[1]
+            self.db.save(self.current_task)
+
+        self.cleanup_env()
+        exit(0)
+
+    def setup_handler(self):
+        """
+        Method to set up the handler in the run method with lower redundancy
+        """
+        log.info('Setting up signal handlers')
+        signal.signal(signal.SIGTERM, self.handler)
+        signal.signal(signal.SIGINT, self.handler)
+
     def prepare_env(self, *args, **kwargs):
-        """Method to be called to prepare the environment to run the
+        """
+        Method to be called to prepare the environment to run the
         application.
         """
 
     def prepare_run(self, *args, **kwargs):
-        """Code to run before a task gets processed. Used e.g. for fetching
+        """
+        Code to run before a task gets processed. Used e.g. for fetching
         inputs.
         """
 
     def process_task(self, task):
-        """The function to override, which processes the tasks themselves.
+        """
+        The function to override, which processes the tasks themselves.
         @param task: the task to process
         """
         raise NotImplementedError
 
     def cleanup_run(self, *args, **kwargs):
-        """Code to run after a task has been processed.
+        """
+        Code to run after a task has been processed.
         """
 
     def cleanup_env(self, *args, **kwargs):
-        """Method which gets called after the run method has completed.
         """
+        Method which gets called after the run method has completed.
+        """
+
+        
+class RunActorWithStop(RunActor):
+    """
+    RunActor class with added stopping functionality.
+    """
+
+    def run(self, max_time=None, avg_time_factor=0.0, max_tasks=0, stop_function=None, **stop_function_args):
```

Comment:
As mentioned earlier, I find it confusing to have `stop_function` as input argument here, if the only use is to allow setting a maximum execution time. This should be done through `max_time`